test/cpp_api_parity/module_impl_check.py

```diff
@@ -0,0 +1,289 @@
+            "If you are adding a new test, please set `fullname` using format `ModuleName_desc`, " + \
+            "or set `module_name` using format `ModuleName`. " + \
+            "(Discovered while processing {}.)".format(test_params_dict)
+
+        module_full_name = 'torch::nn::' + module_name
+
+        assert module_full_name in parity_table['torch::nn'], \
+            "Please add `{}` entry to `torch::nn` section of `test/cpp_api_parity/parity-tracker.md`. (Discovered while processing {}.)".format(
+                module_full_name, test_params_dict)
+
+        for device in devices:
+            test_params = process_test_params_for_module(
+                test_params_dict=test_params_dict,
+                device=device,
+                test_instance_class=test_instance_class,
+            )
+            test_name = 'test_torch_nn_{}'.format(test_params.module_variant_name)
+            torch_nn_test_params_map[test_name] = test_params
+
+            def test_fn(self):
+                test_torch_nn_module_variant(unit_test_class=self, test_params=torch_nn_test_params_map[self._testMethodName])
+
+            test_fn = decorate_test_fn(
+                test_fn=test_fn,
+                test_cpp_api_parity=test_params_dict.get('test_cpp_api_parity', True),
+                test_cuda=test_params_dict.get('test_cuda', True),
+                has_impl_parity=parity_table['torch::nn'][module_full_name][0] and test_params_dict.get('has_parity', True),
+                device=device)
+
+            add_test(unit_test_class, test_name, test_fn)
+
+def add_tests(unit_test_class, test_params_dicts, test_instance_class, parity_table, devices):
+    add_torch_nn_module_impl_parity_tests(
+        parity_table=parity_table,
+        unit_test_class=unit_test_class,
+        test_params_dicts=test_params_dicts,
+        test_instance_class=test_instance_class,
+        devices=devices)
+
+def generate_test_cpp_sources(test_params, template):
+    device = test_params.device
+
+    cpp_constructor_args = test_params.cpp_constructor_args
+    if cpp_constructor_args != '':
+        cpp_constructor_args = '({})'.format(cpp_constructor_args)
+
+    cpp_args_construction_stmts, cpp_forward_args_symbols = compute_cpp_args_construction_stmts_and_forward_arg_symbols(test_params)
+
+    test_cpp_sources = template.substitute(
+        module_variant_name=test_params.module_variant_name,
+        module_qualified_name='torch::nn::{}'.format(test_params.module_name),
+        cpp_args_construction_stmts=";\n  ".join(cpp_args_construction_stmts),
+        cpp_constructor_args=cpp_constructor_args,
+        cpp_forward_args_symbols=", ".join(cpp_forward_args_symbols),
+        cpp_tmp_folder=test_params.cpp_tmp_folder,
+        device=device,
+    )
+    return test_cpp_sources
+
+# Build all C++ tests together, instead of once per test.
+def build_cpp_tests(unit_test_class, print_cpp_source=False):
+    if len(torch_nn_test_params_map) > 0:
```

Comment:
When would this be equal to 0?

Reply:
It should never be equal to 0 - I changed it to assert :D